reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,204 @@
+   The upscaling speed configuration, see explanation in [here](https://docs.ray.io/en/latest/cluster/vms/user-guides/configuring-autoscaling.html#upscaling-and-downscaling-speed).
+ - autoscale_idle_timeout_minutes: Optional[float] (default 1.0)
+   The downscaling speed configuration, see explanation in [here](https://docs.ray.io/en/latest/cluster/vms/user-guides/configuring-autoscaling.html#upscaling-and-downscaling-speed).
+ 
+
+### Implementation
+
+To start a Ray cluster that enables autoscaling, firstly, prepares a config file like following:
+
+```yaml
+cluster_name: spark
+max_workers: 8  # maximum number of workers
+provider:
+    # This is the key of registered Ray autoscaling `NodeProvider` class that is
+    # used as the backend of launching Ray worker nodes.
+    type: spark
+    # This must be true since the nodes share the same ip!
+    use_node_id_as_ip: True
+    disable_node_updaters: True
+    disable_launch_config_check: True
+available_node_types:
+    ray.head.default:  # Ray head configurations
+        resources:
+            CPU: 4
+            GPU: 0
+        node_config: {}
+        max_workers: 0
+    ray.worker:  # Ray worker configurations
+        resources:
+            CPU: 4
+            GPU: 2
+            memory: 2000000000
+            object_store_memory: 1000000000
+        node_config: {}
+        min_workers: 0  # minimum number of workers in this group
+        max_workers: 4  # maximum number of workers in this group
+head_node_type: ray.head.default
+upscaling_speed: 1.0
+idle_timeout_minutes: 0.1
+```
+
+Then, starts a Ray head node via following command (some options are omitted):
+
+```shell
+ray start --head --autoscaling-config=/path/to/autoscaler-config.yaml
+```
+
+The above command starts a Ray head node, it also starts a Ray autoscaler with
+above YAML configuration, it configures autoscaler with "spark" `NodeProvider`
+backend. When Ray cluster is requested to scale up,
+the configured `NodeProvider` is called to trigger Ray worker node setup.
+
+This is `NodeProvider` [interfaces](https://github.com/ray-project/ray/blob/e1c48ab8b1225daa44ea63ed1e7dc956b0e7f411/python/ray/autoscaler/node_provider.py#L13).
+
+### How to implement spark node provider?
+
+For each Ray worker node, we start a spark job with only one spark
+task to hold this Ray worker node, and we set a unique spark job group ID to
+this spark job. When we need to terminate this Ray worker node, we cancel
+the corresponding spark job by canceling corresponding spark job group, so that
+the spark job and its spark task are killed, then it triggers the Ray worker node termination.
+
+Because we can only cancel spark job not spark task when we need to scale down a Ray worker node,
+we have to have one spark job for each Ray worker node.
+
+One thing critical is that spark node provider runs in autoscaler process that is different process
+than the one that executes "setup_ray_cluster" API. User calls "setup_ray_cluster" in
```

Comment:
is it possible to run this spark application *in* the autoscaler node provider process?

Reply:
For databricks user, we can't,

because when we create a databricks notebook, once the databricks notebook is attached to a spark cluster, then spark session is created in the notebook REPL, and user have to call `setup_ray_cluster` API in the notebook REPL, and Ray autoscaler have to reuse the existing spark session.